web-api/src/storage/postgres.rs

```diff
@@ -284,89 +297,154 @@ impl storage::Interest for Storage {
+        Ok(())
+    }
 
-        // bit casting to signed int is fine as we fetch them as signed int before bit casting them back to unsigned int
-        // truncating to 64bit is fine as >292e+6 years is more then enough for this use-case
-        #[allow(clippy::cast_possible_wrap, clippy::cast_possible_truncation)]
-        sqlx::query(
-            "INSERT INTO center_of_interest (coi_id, user_id, is_positive, embedding, view_count, view_time_ms, last_view)
-            VALUES ($1, $2, $3, $4, $5, $6, $7)
-            ON CONFLICT (coi_id) DO UPDATE SET
-                embedding = EXCLUDED.embedding,
-                view_count = EXCLUDED.view_count,
-                view_time_ms = EXCLUDED.view_time_ms,
-                last_view = EXCLUDED.last_view;",
-        )
-        .bind(updated_coi.id.as_ref())
-        .bind(user_id)
-        .bind(true)
-        .bind(updated_coi.point.to_vec())
-        .bind(updated_coi.stats.view_count as i32)
-        .bind(updated_coi.stats.view_time.as_millis() as i64)
-        .bind(timestamp)
-        .execute(&mut tx)
-        .await?;
+    async fn upsert_tag_weights<'c, I>(
+        tx: &mut Transaction<'_, Postgres>,
+        user_id: &UserId,
+        updates: I,
+    ) -> Result<(), Error>
+    where
+        I: IntoIterator<Item = (&'c str, i32)>,
+    {
+        let updates = updates.into_iter();
 
-        sqlx::query(
-            "INSERT INTO interaction (doc_id, user_id, time_stamp, user_reaction)
-            VALUES ($1, $2, $3, $4)
-            ON CONFLICT (doc_id, user_id, time_stamp) DO UPDATE SET
-                user_reaction = EXCLUDED.user_reaction;",
-        )
-        .bind(doc_id)
-        .bind(user_id)
-        .bind(timestamp)
-        .bind(UserInteractionType::Positive as i16)
-        .execute(&mut tx)
-        .await?;
+        let mut builder = QueryBuilder::new("INSERT INTO weighted_tag (user_id, tag, weight)");
+        let mut iter = updates.into_iter().peekable();
+        while iter.peek().is_some() {
+            let chunk = (&mut iter).take(Database::BIND_LIMIT / 7);
+            builder
+                .reset()
+                .push_values(chunk, |mut builder, (category, weight_diff)| {
+                    builder
+                        .push_bind(user_id)
+                        .push_bind(category)
+                        .push_bind(weight_diff);
+                })
+                .push(
+                    "ON CONFLICT (user_id, category) DO UPDATE SET
+                    weight = weighted_category.weight + EXCLUDED.weight;",
+                )
+                .build()
+                .persistent(false)
+                .execute(&mut *tx)
+                .await?;
+        }
+        Ok(())
+    }
+}
 
-        tx.commit().await?;
+#[derive(FromRow)]
+struct QueriedCoi {
+    coi_id: Uuid,
+    is_positive: bool,
+    embedding: Vec<f32>,
+    /// The count is a `usize` stored as `i32` in database
```

Comment:
:pushpin: Know that this is not really a change from this PR. We need a following PR to not have `count` as `usize` since has nothing to do with the pointer to memory, it should be `u32` or `u64`. I'm afraid of what it could happen if we change architecture.